tests/csrf_protection_test.py

```diff
@@ -0,0 +1,361 @@
+    """Obtains a CSRF token by accessing the signup page, then submits the signup
+    form with valid credentials and the retrieved CSRF token
+    to verify successful submission.
+
+    Args:
+        client: The test client for the application.
+
+    Returns:
+        None
+    """
+    response = client.get("/signup")
+    csrf_token = get_csrf_token_from_response(response.data.decode())
+
+    response = client.post(
+        "/signup",
+        data={
+            "email": "test@example.com",
+            "password": "TestP@ss35!@!@#daf3",
+            "password_confirm": "TestP@ss35!@!@#daf3",
+            "csrf_token": csrf_token,
+        },
+        follow_redirects=True,
+    )
+    assert (
+        response.status_code == 200
+    ), "Form should accept submission with valid CSRF token"
+
+
+def test_signup_post_invalid_csrf(client):
+    """Attempts to submit the signup form with an invalid CSRF
+    token to verify that the form correctly rejects the submission.
+
+    Args:
+        client: The test client for the application.
+
+    Returns:
+        None
+    """
+    response = client.post(
+        "/signup",
+        data={
+            "email": "test@example.com",
+            "password": "T$R%#@$FASD3m2o3sdf",
+            "password_confirm": "T$R%#@$FASD3m2o3sdf",
+            "csrf_token": "invalid_csrf_token",
+        },
+        follow_redirects=True,
+    )
+    assert (
+        "Invalid CSRF token. Please try again." in response.data.decode()
+        or response.status_code == 400
+    ), "Form should reject submission with invalid CSRF token"
+
+
+def test_signup_post_valid_csrf(client):
+    """Submits the signup form with a valid CSRF token to verify
+    that the form accepts the submission.
+
+    Args:
+        client: The test client for the application.
+
+    Returns:
+        None
+    """
+    response = client.get("/signup")
+    csrf_token = get_csrf_token_from_response(response.data.decode())
+
+    response = client.post(
+        "/signup",
+        data={
+            "email": "test@example.com",
+            "password": "T$R%#@$3o2m3v23lkmasdf",
+            "password_confirm": "T$R%#@$3o2m3v23lkmasdf",
+            "csrf_token": csrf_token,
+        },
+        follow_redirects=True,
+    )
+
+
+def test_change_password_form_rejection_without_csrf(client):
+    """Verify that the change password form does not process
+    submissions that lack a CSRF token.
+
+    Args:
+        client: The test client for the application.
+
+    Returns:
+        None
+    """
+    response = client.post(
+        "/change_password",
+        data={
+            "old_password": "TestingP@ss35",
+            "new_password": "NewTestingPsdf23r55",
+            "new_password_confirm": "TestP@NewTestingPsdf23r55",
+        },
+        follow_redirects=True,
+    )
+    assert (
+        "The CSRF token is missing." in response.data.decode()
+        or response.status_code == 400
```

Comment:
Check this link: https://docs.python.org/3/library/http.html

we can use:

`response.status_code == HTTPStatus.BAD_REQUEST`